### PR TITLE
[AOTI] Improve the two-pass wrapper codegen

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113762

Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change removes kernel scheduling inconsistency between the two passes and also speeds up the compilation. Another future improvement can be done is to make the second-pass reuse the scheduled nodes and do the wrapper codegen only.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D51405702](https://our.internmc.facebook.com/intern/diff/D51405702)

### CODE REVIEW

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51405702).

File: test/inductor/test_aot_inductor.py [hunk 1]
```diff
@@ -1177,6 +1214,7 @@ class AOTInductorTestABICompatibleCpu(TestCase):
         # TODO: test_freezing_abi_compatible_cpu somehow fails on CI but not locally,
         #   NotImplementedError: Cannot access storage of OpaqueTensorImpl
         "test_freezing": TestFailure(("abi_compatible_cpu",), is_skip=True),
+        "test_missing_cubin": TestFailure(("abi_compatible_cpu",)),
```
Comment (muchulee8) [inline test/inductor/test_aot_inductor.py:1217]: What will happen if we run cpu tests? It is true that the main purpose is to store/load cubin in different stage, but in theory it should still work fine if we run the exact CPU test?

File: test/inductor/test_aot_inductor.py [hunk 2]
```diff
@@ -1202,6 +1240,7 @@ class AOTInductorTestABICompatibleCuda(TestCase):
     # test_failures, xfail by default, set is_skip=True to skip
     {
         "test_dup_unbacked_sym_decl": TestFailure(("abi_compatible_cuda",)),
+        "test_missing_cubin": TestFailure(("abi_compatible_cuda",)),
```
Comment (muchulee8) [inline test/inductor/test_aot_inductor.py:1243]: Can we add a comment on what happened? The `# There is a double-free issue which will be fixed in another PR` at the bottom is pretty informative, and people know it will be fixed. It's hard to tell for other tests whether the XFAIL is in purpose or it's need to be fixed?

File: torch/_inductor/graph.py [hunk 3]
```diff
@@ -974,10 +1019,11 @@ def compile_to_module(self):
         return mod
 
     def compile_to_fn(self):
-        if self.aot_mode and self.cpp_wrapper:
+        if self.aot_mode:
             from .codecache import AotCodeCache
 
-            code, linemap = self.codegen()
+            assert self.cpp_wrapper, "AOT mode only supports C++ wrapper"
```
Comment (muchulee8) [inline torch/_inductor/graph.py:1025]: Nice!

File: test/inductor/test_aot_inductor.py (hunk 1 above)
Comment (desertfire) [reply test/inductor/test_aot_inductor.py:1217]: We need to add abi_compatible support for convolution, and the issue is more than adding a missing shim interface. I will add a comment, and come back to fix it later.

File: torch/_inductor/compile_fx.py [hunk 4]
```diff
@@ -590,8 +516,11 @@ def fx_codegen_and_compile(
         post_grad_graphs_log.info("%s", lazy_format_graph_code("AFTER POST GRAD", gm))
 
     with V.set_fake_mode(fake_mode):
+        # example_inputs will be used by AOTInductor to dry-run the generated code for Triton kernel tuning,
+        # and for the forward pass, we have the real inputs to be used as example_inputs.
         graph = GraphLowering(
             gm,
+            example_inputs=V.real_inputs if is_inference else example_inputs,
```
Comment (chenyang78) [inline torch/_inductor/compile_fx.py:523]: nit - Looks like for the non-forward pass, ```example_inputs``` are faked tensors or sym values. Could you add a comment to make it clear? Sometimes, it's hard to tell if an example_input is of FakedTensor or not by reading the code. Thanks.

**Review by chenyang78 — APPROVED** [review]
Left a minor comment. LGTM, Thanks!

**Review by muchulee8 — APPROVED** [review]
